triedb/pathdb/database.go

```diff
@@ -0,0 +1,398 @@
+	readOnly   bool                // Flag if database is opened in read only mode
+	bufferSize int                 // Memory allowance (in bytes) for caching dirty nodes
+	config     *Config             // Configuration for database
+	diskdb     ethdb.KeyValueStore // Persistent storage for matured trie nodes
+	tree       *layerTree          // The group for all known layers
+	lock       sync.RWMutex        // Lock to prevent mutations from happening at the same time
+	dirties    dbtypes.KvMap
+}
+
+// New attempts to load an already existing layer from a persistent key-value
+// store (with a number of memory layers from a journal). If the journal is not
+// matched with the base persistent layer, all the recorded diff layers are discarded.
+func New(diskdb ethdb.KeyValueStore, config *Config) *Database {
+	if config == nil {
+		config = Defaults
+	}
+	config = config.sanitize()
+	db := &Database{
+		readOnly:   config.ReadOnly,
+		bufferSize: config.DirtyCacheSize,
+		config:     config,
+		diskdb:     diskdb,
+		dirties:    make(dbtypes.KvMap),
+	}
+	// Construct the layer tree by resolving the in-disk singleton state
+	// and in-memory layer journal.
+	db.tree = newLayerTree(db.loadLayers())
+
+	return db
+}
+
+// Reader retrieves a layer belonging to the given state root.
+func (db *Database) Reader(root common.Hash) (layer, error) {
+	l := db.tree.get(root)
+	if l == nil {
+		return nil, fmt.Errorf("state %#x is not available", root)
+	}
+	return l, nil
+}
+
+func (db *Database) CommitGenesis(root common.Hash) error {
+	log.Info("pathdb write genesis state to disk", "root", root.Hex())
+	batch := db.diskdb.NewBatch()
+	for _, v := range db.dirties {
+		batch.Put(v.K, v.V)
+	}
+	for k := range db.dirties {
+		delete(db.dirties, k)
+	}
+	if err := batch.Write(); err != nil {
+		return err
+	}
+	batch.Reset()
+
+	// Update stateID
+	rawdb.WriteStateID(db.diskdb, root, 0)
+	return nil
+}
+
+// Commit traverses downwards the layer tree from a specified layer with the
+// provided state root and all the layers below are flattened downwards. It
+// can be used alone and mostly for test purposes.
+func (db *Database) CommitState(root common.Hash, parentRoot common.Hash, blockNumber uint64, report bool) error {
+	// Hold the lock to prevent concurrent mutations.
+	db.lock.Lock()
+	defer db.lock.Unlock()
+
+	// Short circuit if the mutation is not allowed.
+	if err := db.modifyAllowed(); err != nil {
+		return err
+	}
+
+	// only 1 entity, state have no changes
+	// some block maybe has no txns, so state do not change
+	if root == parentRoot && len(db.dirties) == 1 {
+		return nil
+	}
+
+	nodes := db.dirties.Flatten()
+	for k := range db.dirties {
+		delete(db.dirties, k)
+	}
+
+	clearNodes := func() {
+		// clear tmp nodes
+		for k := range nodes {
+			delete(nodes, k)
+		}
+	}
+
+	if err := db.tree.add(root, parentRoot, blockNumber, nodes); err != nil {
+		clearNodes()
+		return err
+	}
+
+	clearNodes()
```

Comment:
Instead of using `Flatten()` which creates a copy, you can directly pass `db.dirties` to `db.tree.add` and then reset `db.dirties` by assigning it a new empty map. This avoids unnecessary copying and improves efficiency. e.g.

```
if err := db.tree.add(root, parentRoot, blockNumber, db.dirties); err != nil {
    return err
}

// Reset db.dirties
db.dirties = make(KvMap)
```